src/tribol/geom/NodalNormal.hpp

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and
+// other Tribol Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (MIT)
+
+#ifndef SRC_GEOM_NODALNORMAL_HPP_
+#define SRC_GEOM_NODALNORMAL_HPP_
+
+#include "tribol/config.hpp"
+
+#include "tribol/mesh/MeshData.hpp"
+#include "tribol/mesh/MethodCouplingData.hpp"
+
+namespace tribol {
+
+/**
+ * @brief Virtual base class to define the interface for nodal normal calculations
+ */
+class NodalNormal {
+ public:
+  /**
+   * @brief Destructor
+   */
+  virtual ~NodalNormal() {}
+
+  /**
+   * @brief Interface for computing and storing nodal normals on a given mesh
+   *
+   * @param mesh Mesh data
+   */
+  virtual void Compute( MeshData& mesh ) = 0;
+
+  /**
+   * @brief Gets the Jacobian data for the nodal normals
+   *
+   * @pre Must call NodalNormal::Compute() first
+   *
+   * @return MethodData reference storing Jacobian data
+   */
+  MethodData& getJacobianData() { return elem_jacobians_; }
+
+  /// @overload
+  const MethodData& getJacobianData() const { return elem_jacobians_; }
+
+ private:
+  /**
+   * @brief Stores Jacobian contributions of the nodal normals
+   */
+  MethodData elem_jacobians_;
+};
+
+/**
+ * @brief Computes nodal normals as the average of connected element normals
+ */
+class ElementAvgNormal : public NodalNormal {
+ public:
+  /**
+   * @brief Computes nodal normals as the average of connected element normals
+   *
+   * @param mesh Mesh data
+   */
+  void Compute( MeshData& mesh ) override;
+};
+
+#ifdef TRIBOL_USE_ENZYME
+
+/**
+ * @brief Computes nodal normals by averaging normal evaluated at the node of all connected elements
```

Comment:
I think this brief needs to be elaborated. I'm not following how this normal is computed. Is this the approach where each edge that shares a given node is used to compute a nodal normal (using that edge) and those edge normals are then averaged? Or does this use a continuously defined normal on each face that shares a node and then evaluates that normal (function) at the node of interest? Seems like the latter is not the case since we only have face/element normals at discrete points? If it is the former, perhaps we call this the EdgeAvgNodalNormal?

Reply:
EdgeAvgNodalNormal ?

Reply:
changed